include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +853,207 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+			{
+				localH = irr_glsl_beckmann_cos_generate_wo_clamps(upperHemisphereLocalV, u.xy, ax, ay);
+			} CASE_END
 #endif //NDF_PHONG
-	{}
+			CASE_OTHERWISE
+			{}
+			END_CASES
+
+			vec3 localL;
+			const float VdotH = dot(localH, localV);
+			const float VdotH_clamp = is_bsdf ? VdotH : max(VdotH, 0.0);
+			vec3 fr;
+			bool refraction = false;
+#ifdef OP_CONDUCTOR
+			if (op == OP_CONDUCTOR)
+			{
+				fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+				rem *= fr;
+			}
+			else
+#endif
+			{
+				fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
+
+				const float refractionProb = colorToScalar(fr);
+				float rcpChoiceProb;
+				refraction = irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb);
+				localPdf /= rcpChoiceProb;
+			}
+			float eta = colorToScalar(ior[0]);
+			float rcpEta = 1.0/eta;
+
+			out_microfacet.inner = irr_glsl_calcAnisotropicMicrofacetCache(refraction, localV, localH, localL, rcpEta, rcpEta*rcpEta);
+			s = irr_glsl_createLightSampleTangentSpace(localV, localL, tangentFrame);
+
+			finalizeMicrofacet(out_microfacet);
+			const float TdotH2 = out_microfacet.TdotH2;
+			const float BdotH2 = out_microfacet.BdotH2;
+			const float NdotH2 = out_microfacet.inner.isotropic.NdotH2;
+			const float NdotL = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+
+			BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+			CASE_BEGIN(ndf, NDF_GGX)
+			{
+				G2_over_G1 = irr_glsl_ggx_smith_G2_over_G1(NdotL, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+				G1_over_2NdotV = irr_glsl_GGXSmith_G1_wo_numerator(NdotV, TdotV2, BdotV2, NdotV2, ax2, ay2);
+				ndf_val = irr_glsl_ggx_aniso(TdotH2, BdotH2, NdotH2, ax, ay, ax2, ay2);
+			} CASE_END
+#endif //NDF_GGX
+
+#ifdef NDF_BECKMANN
+			CASE_BEGIN(ndf, NDF_BECKMANN)
+			{
+				float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+				G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+				G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, ax2, ay2);
+				ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+			} CASE_END
+#endif //NDF_BECKMANN
+
+#ifdef NDF_PHONG
+			CASE_BEGIN(ndf, NDF_PHONG)
+			{
+				float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+				G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+				G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL * s.TdotL, s.BdotL * s.BdotL, s.NdotL * s.NdotL, ax2, ay2);
+				ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+			} CASE_END
+#endif //NDF_PHONG
+			CASE_OTHERWISE
+			{}
+			END_CASES
+
+			const float LdotH = out_microfacet.inner.isotropic.LdotH;
+			const float VdotHLdotH = VdotH * LdotH;
+			// a trick, pdf was already multiplied by transmission/reflection choice probability above
+			const float reflectance = refraction ? 0.0 : 1.0;
```

Comment:
then make a specialized `irr_glsl_smith_VNDF_pdf_wo_clamps` that transforms a `FVNDF`  and does not take a `refraction` parameter ;)

Reply:
```glsl
float irr_glsl_smith_FVNDF_pdf_wo_clamps(in float fresnel_ndf, in float G1_over_2NdotV, in float absNdotV, in bool transmitted, in float VdotH, in float LdotH, in float VdotHLdotH, in float orientedEta)
{
    float FNG = fresnel_ndf*G1_over_2NdotV;
    float factor = 0.5;
    if (transmitted)
    {
        const float VdotH_etaLdotH = (VdotH+orientedEta*LdotH);
        // VdotHLdotH is negative under transmission, so this factor is negative
        factor *= -2.0*VdotHLdotH/(VdotH_etaLdotH*VdotH_etaLdotH);
    }
    return FNG*factor;
}
```

and then rewrite https://github.com/Devsh-Graphics-Programming/Nabla/blob/shader_pipeline_irrlichtbaw/include/irr/builtin/glsl/bxdf/geom/smith/common.glsl#L33

in terms of the new thing